send-transaction-service/src/send_transaction_service.rs

```diff
@@ -137,49 +169,66 @@ impl SendTransactionService {
         let mut transactions = HashMap::new();
 
+        info!(
+            "Starting send-transaction-service thread with config {:?}",
+            config
+        );
         if let Some(leader_info) = leader_info.as_mut() {
             leader_info.refresh_recent_peers();
         }
         connection_cache::set_use_quic(config.use_quic);
         Builder::new()
             .name("send-tx-sv2".to_string())
             .spawn(move || loop {
-                match receiver.recv_timeout(Duration::from_millis(1000.min(config.retry_rate_ms))) {
+                let mut do_batch = false;
+                let recv_timeout_ms = config.retry_rate_ms.min(config.batch_send_rate_ms);
+                match receiver.recv_timeout(Duration::from_millis(1000.min(recv_timeout_ms))) {
                     Err(RecvTimeoutError::Disconnected) => break,
-                    Err(RecvTimeoutError::Timeout) => {}
+                    Err(RecvTimeoutError::Timeout) => {
+                        if !transactions.is_empty() {
+                            // Timed out waiting for the full batch and there is something in the batch
+                            inc_new_counter_info!("send_transaction_service-batch-time-out-tx", 1);
+                            do_batch = true;
+                        }
+                    }
                     Ok(transaction_info) => {
                         inc_new_counter_info!("send_transaction_service-recv-tx", 1);
-                        let transactions_len = transactions.len();
+                        let mut transactions_len = transactions.len();
                         let entry = transactions.entry(transaction_info.signature);
                         if let Entry::Vacant(_) = entry {
-                            let addresses = leader_info.as_ref().map(|leader_info| {
-                                leader_info.get_leader_tpus(config.leader_forward_count)
-                            });
-                            let addresses = addresses
-                                .map(|address_list| {
-                                    if address_list.is_empty() {
-                                        vec![&tpu_address]
-                                    } else {
-                                        address_list
-                                    }
-                                })
-                                .unwrap_or_else(|| vec![&tpu_address]);
-                            for address in addresses {
-                                Self::send_transaction(address, &transaction_info.wire_transaction);
-                            }
-                            if transactions_len < MAX_TRANSACTION_QUEUE_SIZE {
-                                inc_new_counter_info!("send_transaction_service-insert-tx", 1);
-                                entry.or_insert(transaction_info);
-                            } else {
+                            if transactions_len >= MAX_TRANSACTION_QUEUE_SIZE {
```

Comment:
I think the better behavior here is to actually append this transaction to the batch then send the entire batch. If the queue is already full, we presumably don't have too few transactions, so it would make sense to flush the queue at this point and send the entire queue. And in the interest if efficiency, it makes sense to append the current transaction to the queue and send the queue rather than send the current transaction individually then send the queue. This doe technically make the queue larger than MAX_TRANSACTION_QUEUE_SIZE for a brief time (although we immediately send the queue anyway), so to make this more strictly correct, the condition can be changed to 

if transactions_len >= MAX_TRANSACTION_QUEUE_SIZE - 1

Reply:
The queue has dual purposes -- doing retries for transactions submitted earlier and saving the entries for batching. This is logic is a stop gag method, which is not newly introduced. I think it is safer to keep this logic.